ufl/tensors.py

```diff
@@ -235,14 +235,24 @@ def _simplify_indexed(self, multiindex):
         """Return a simplified Expr used in the constructor of Indexed(self, multiindex)."""
         # Untangle as_tensor(C[kk], jj)[ii] -> C[ll]
         B, jj = self.ufl_operands
+        if len(multiindex) != len(jj):
+            raise ValueError(f"len(multiindex) ({len(multiindex)}) != len(jj) ({len(jj)})")
+        rep = dict(zip(jj, multiindex))
+        # Avoid recursion and just attempt to simplify some common patterns
+        # as the result of this method is not cached.
+        if isinstance(B, Indexed):
+            C, kk = B.ufl_operands
+            if isinstance(C, ListTensor) and len(kk) == 1 and isinstance(rep[kk[0]], FixedIndex):
+                (k,) = kk
+                B = C.ufl_operands[int(rep[k])]
+                jj = MultiIndex(tuple(j for j in jj if j != k))
+                multiindex = MultiIndex(tuple(rep[j] for j in jj if j != k))
```

Comment:
The index `k` should no longer be in the modified `jj`
```suggestion
                multiindex = MultiIndex(tuple(rep[j] for j in jj))
```